components/list/list-item-mixin.js

```diff
@@ -523,6 +544,11 @@ export const ListItemMixin = superclass => class extends composeMixins(
 		}
 	}
 
+	_handleButtonAddClick() {
+		/** Dispatched when the add button directly after the item is clicked */
+		this.dispatchEvent(new CustomEvent('d2l-list-item-add-button-click', { bubbles: true, composed: true }));
```

Comment:
Hmmm, I'm sort of undecided on this at the moment, but these are my initial thoughts...

The consumer is going to need to know which add-button was clicked. If the event is dispatched from the item, then they will know the reference item via `e.target`. If we dispatch the event from the list, how will they know? - I think we'd need to provide them with event detail such as the item `key`, which is currently at least defined and used by `ListItemCheckboxMixin` and `ListItemDragDropMixin`.

From an ergonomics perspective, if the event is defined to bubble from the item (which it currently is), technically consumers could wire-up a single handler to the list.

But should it bubble? We should consider nested lists where a use clicks an add-button in a nested list. Currently, it looks like the event will bubble up and event handlers on ancestor lists would be invoked. Is that what we'd want? It seems like that might make the consumer's ancestor handler more error prone since it would have to consider that the event might not be coming from its own items.

If the event handler was wired up to the item, and that handler is defined in their component with references to their list state, ids, etc., but then the item is moved to a different list. The original handler might be holding reference to stale state. However, I think our drag and drop creates brand new items when items are moved, and the original DOM not is removed, so maybe that would not be a problem in practice. Nevertheless, something to think about, and kinda leads me more to thinking the event belongs more to the list itself, rather than the items.

The event is currently set to `composed: true`. Do we need it to traverse DOM scopes?

Reply:
Great points.

I would say I agree the in the nested list case we actually would not want it to bubble. It does sound safest to add the listener to the list itself and use the keys for identifying the list item whose button was clicked. Then everything `add-button` would be added in the same place as well. I don't think we need the event to traverse DOM scopes.

Reply:
Ok. Sounds good. I notice the code still dispatches `d2l-list-item-add-button-click` from the list-item. Are you planning to change it so that the event is dispatched from `d2l-list`, and configured to not bubble?

Reply:
So I set it up so that the event chain goes: `click` event (`d2l-button-add`) > `d2l-list-item-add-button-click` event (`list-item-mixin`) > `d2l-list-add-button-click` event (`list`). I noticed that if I didn't have `bubbles` on the list item event then the list event listener for `d2l-list-item-add-button-click` didn't actually get triggered. Is there something wrong with my setup for this?

Reply:
Just that in the nested case, higher level lists would also get the event. If we don't want those higher level lists to see the events coming from nested lists, we'd either have to:

a) dispatch the `d2l-list-add-button-click` event directly from the `d2l-list` element and use `bubbles: false`, or
b) `d2l-list` would need to implement an event handler for it that calls `e.stopPropagation();` so that the event doesn't bubble any further than the immediate list parent. In this case, hopefully `stopPropagation` would not prevent _all_ the handlers (ours as well as the consumers) on `d2l-list` from running.

Reply:
Ah gotcha - I did add `e.stopPropogation` to the `list` handler for the list item event but it sounds like that might be riskier so I can switch to method (a) if that's preferred

Reply:
Ah sorry. I didn't notice that. I think either way if fine. If we run into any problem with the consumer's handler not running, we can switch.